eppo_core/src/event_ingestion/event_ingestion.rs

```diff
@@ -0,0 +1,228 @@
+
+impl EventIngestionConfig {
+    const DEFAULT_BATCH_SIZE: usize = 1_000;
+    const DEFAULT_DELIVERY_INTERVAL: Duration = Duration::from_secs(10);
+    const DEFAULT_RETRY_INTERVAL: Duration = Duration::from_secs(5);
+    const DEFAULT_MAX_RETRY_DELAY: Duration = Duration::from_secs(30);
+    const DEFAULT_MAX_RETRIES: u32 = 3;
+    const DEFAULT_MAX_QUEUE_SIZE: usize = 10_000;
+
+    /// Creates new event ingestion config.
+    ///
+    /// Returns `None` if `sdk_key` is not suitable for event ingestion.
+    pub fn new(sdk_key: SdkKey) -> Option<Self> {
+        let ingestion_url = sdk_key.event_ingestion_url()?;
+        let config = EventIngestionConfig {
+            sdk_key,
+            ingestion_url,
+            delivery_interval: Self::DEFAULT_DELIVERY_INTERVAL,
+            retry_interval: Self::DEFAULT_RETRY_INTERVAL,
+            max_retry_delay: Self::DEFAULT_MAX_RETRY_DELAY,
+            max_retries: Self::DEFAULT_MAX_RETRIES,
+            batch_size: Self::DEFAULT_BATCH_SIZE,
+            max_queue_size: Self::DEFAULT_MAX_QUEUE_SIZE,
+        };
+        Some(config)
+    }
+
+    pub fn spawn(&self, runtime: &BackgroundRuntime) -> EventIngestion {
+        EventIngestion::spawn(runtime, self)
+    }
+}
+
+/// A handle to Event Ingestion subsystem.
+pub struct EventIngestion {
+    tx: mpsc::Sender<BatchedMessage<QueuedEvent>>,
+}
+
+impl EventIngestion {
+    /// Starts the event ingestion subsystem on the given background runtime.
+    pub fn spawn(runtime: &BackgroundRuntime, config: &EventIngestionConfig) -> EventIngestion {
+        let event_delivery =
+            EventDelivery::new(config.sdk_key.clone(), config.ingestion_url.clone());
+
+        let (input, flusher_uplink) = mpsc::channel(config.max_queue_size);
+        let (flusher_downlink, batcher_uplink) = mpsc::channel(1);
+        let (batcher_downlink, delivery_uplink) = mpsc::channel(1);
+        let (delivery_downlink, retry_uplink) = mpsc::channel(1);
+        let (delivery_status_tx, delivery_status_rx) = mpsc::channel(1);
+
+        runtime.spawn_untracked(auto_flusher::auto_flusher(
+            flusher_uplink,
+            flusher_downlink,
+            config.delivery_interval,
+        ));
+        runtime.spawn_untracked(batcher::batcher(
+            batcher_uplink,
+            batcher_downlink.clone(),
+            config.batch_size,
+        ));
+        runtime.spawn_untracked(delivery::delivery(
+            delivery_uplink,
+            delivery_downlink,
+            delivery_status_tx.clone(),
+            config.max_retries,
+            event_delivery,
+        ));
+        runtime.spawn_untracked(retry::retry(
+            retry_uplink,
+            // This is sending to input of delivery
+            batcher_downlink,
+            config.max_retries,
+            config.retry_interval,
+            config.max_retry_delay,
+        ));
+
+        // For now, nobody is interested in delivery statuses.
+        let _ = delivery_status_rx;
```

Comment:
i was using this in tests to assert on the output of the "pipeline" but if you think the wiremock assertions are sufficient, that works for me